mover-restic/entry.sh

```diff
@@ -72,12 +76,156 @@ function do_prune {
+    local date_as_epoch
+    date_as_epoch=$(date --date="${timestamp}" +%s)
+    echo "${date_as_epoch}"
+}
+
+
+#######################################
+# Reverses the elements within an array,
+# inspired by: https://unix.stackexchange.com/a/412870
+# Globals:
+#   None
+# Arguments:
+#   Name of array
+#######################################
+function reverse_array() {
+    local -n _arr=$1
+
+    # set indices
+    local -i left=0
+    local -i right=$((${#_arr[@]} - 1))
+
+    while ((left < right)); do
+        # triangle swap
+        local -i temp="${_arr[$left]}"
+        _arr[$left]="${_arr[$right]}"
+        _arr[$right]="$temp"
+
+        # increment indices
+        ((left++))
+        ((right--))
+    done
+}
+
+################################################################
+# Selects the first restic snapshot available for the 
+# given constraints. If RESTORE_AS_OF is defined, then 
+# only snapshots that were created prior to it are considered.
+# If SELECT_PREVIOUS is defined, then the n-th snapshot 
+# is selected under the matching criteria. 
+# If a snapshot satisfying the conditions is found, then its ID 
+# is returned.
+# 
+# Globals:
+#   SELECT_PREVIOUS
+#   RESTORE_AS_OF
+# Arguments:
+#   None
+################################################################
+function select_restic_snapshot_to_restore() {
+    # list of epochs
+    declare -a epochs
+    # create an associative array that maps numeric epoch to the restic snapshot IDs
+    declare -A epochs_to_snapshots
+
+    # declare vars to be used in the loop
+    local snapshot_id
+    local snapshot_ts
+    local trimmed_timestamp
+    local snapshot_epoch 
+
+    # go through the timestamps received from restic
+    IFS=$'\n'
+    for line in $(restic -r "${RESTIC_REPOSITORY}" snapshots | grep /data | awk '{print $1 "\t" $2 " " $3}'); do
+        # extract the proper variables
+        snapshot_id=$(echo -e "${line}" | cut -d$'\t' -f1)
+        snapshot_ts=$(echo -e "${line}" | cut -d$'\t' -f2)
+        trimmed_timestamp=$(trim_timestamp "${snapshot_ts}")
+        snapshot_epoch=$(get_epoch_from_timestamp "${trimmed_timestamp}")
+        epochs+=("${snapshot_epoch}")
+        epochs_to_snapshots[${snapshot_epoch}]="${snapshot_id}"
+    done
+
+    # reverse sorted epochs so the most recent epoch is first 
+    reverse_array epochs
+
+    local -i offset=0
+    local target_epoch
+    if [[ -n ${RESTORE_AS_OF} ]]; then
+        # move to the position of the satisfying epoch
+        target_epoch=$(get_epoch_from_timestamp "${RESTORE_AS_OF}")
+        while ((offset < ${#epochs[@]})); do 
+            if ((${epochs[${offset}]} <= target_epoch)); then
+                break
+            fi
+            ((offset++))
+        done
+    fi
+
+    # offset the epoch selection if SELECT_PREVIOUS is defined
+    local -i select_offset=${SELECT_PREVIOUS-0}
+    ((offset+=select_offset))
+
+    # when timestamp isnt provided, just retrieve the latest snapshot
```

Comment:
This comment really confused me. When timestamp isn't provided, offset is 0, so we'll just take the most recent.
This block is really just converting the offset to the snapshot id, right?

Reply:
This comment was forgotten after this function was drastically changed. A Classic example of code debt.